src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs

```diff
@@ -538,7 +538,23 @@ internal int FindEntry(TKey key, out object? value)
             {
                 Debug.Assert(key != null); // Key already validated as non-null.
 
-                int hashCode = RuntimeHelpers.GetHashCode(key) & int.MaxValue;
+                int hashCode;
+
+#if CORECLR || NATIVEAOT
+                hashCode = RuntimeHelpers.TryGetHashCode(key);
+
+                if (hashCode == 0)
+                {
+                    // No hash code has been assigned to the key, so therefore it has not been added
+                    // to any ConditionalWeakTable.
+                    value = null;
+                    return -1;
+                }
+#else
+                hashCode = RuntimeHelpers.GetHashCode(key);
+#endif
```

Comment:
```suggestion
#if MONO
                hashCode = RuntimeHelpers.GetHashCode(key);
#else
                hashCode = RuntimeHelpers.TryGetHashCode(key);

                if (hashCode == 0)
                {
                    // No hash code has been assigned to the key, so therefore it has not been added
                    // to any ConditionalWeakTable.
                    value = null;
                    return -1;
                }
#endif
```
Simpler `#if` condition

Reply:
Would it be easier to just make Mono not create `0` hashcodes?  Like add `if (h == 0) h = 1;` somewhere after the value is computed?

Reply:
I think that would be possible:

https://github.com/AustinWise/runtime/commit/1ccbe6182a55b913c521441f18249c7d61a8cb10

It's a bit more involved and I'm less familiar with testing Mono changes. If this PR get merged before I get a chance to test and benchmark the changes, I'll make a followup PR.

Reply:
I updated the PR with the Mono changes and some benchmarks for Mono on X64 Linux and Mono on WASM.